File: src/ast/EncodedFunction.java
package ast;

import java.io.StringWriter;

public class EncodedFunction extends BaseNode {
    private BaseNode _name;
    private BaseNode _params;
    private BaseNode _cv;
    private BaseNode _ref;
    private BaseNode _attrs;
    private BaseNode _ret;

    public EncodedFunction(BaseNode name, BaseNode Params, BaseNode cv, BaseNode Ref, BaseNode attrs, BaseNode ret) {
        super(NodeType.NameType);
        _name = name;
        _params = Params;
        _cv = cv;
        _ref = Ref;
        _attrs = attrs;
        _ret = ret;
    }

    @Override
    public void PrintLeft(StringWriter writer) {
        if (_ret != null) {
            _ret.PrintLeft(writer);

            if (!_ret.HasRightPart()) {
                writer.write(" ");
            }
        }

        _name.Print(writer);

    }

    @Override
    public boolean HasRightPart() {
        return true;
    }

    @Override
    public void PrintRight(StringWriter writer) {
        writer.write("(");

        if (_params != null) {
            _params.Print(writer);
        }

        writer.write(")");

        if (_ret != null) {
            _ret.PrintRight(writer);
        }

        if (_cv != null) {
            _cv.Print(writer);
        }

        if (_ref != null) {
            _ref.Print(writer);
        }

        if (_attrs != null) {
            _attrs.Print(writer);
        }
    }
}

File: src/de/fearlesstobi/demangler/ast/EncodedFunction.kt
package de.fearlesstobi.demangler.ast

import java.io.StringWriter

class EncodedFunction(val name: BaseNode, private val params: BaseNode?, private val cv: BaseNode?, private val ref: BaseNode?, private val attrs: BaseNode?, private val ret: BaseNode?) : BaseNode(NodeType.NameType) {
    public override fun printLeft(writer: StringWriter) {
        if (ret != null) {
            ret.printLeft(writer)
            if (!ret.hasRightPart()) {
                writer.write(" ")
            }
        }
        name.print(writer)
    }

    public override fun hasRightPart(): Boolean {
        return true
    }

    public override fun printRight(writer: StringWriter) {
        writer.write("(")
        params?.print(writer)
        writer.write(")")
        ret?.printRight(writer)
        cv?.print(writer)
        ref?.print(writer)
        attrs?.print(writer)
    }
}